### DISCUSSION TITLE
Can KeyCloak automatically resolve the realm for the user ?

### DISCUSSION THREAD

**Discussion by voroninp** [keycloak/keycloak#31610] [Q&A]
Sorry if the question is stupid, I am new to KC.

As I can see from [the docs](https://www.keycloak.org/docs/latest/securing_apps/index.html#endpoints), users should always be redirected to a specific realm for authentication.

But can CK infer the realm automatically from entered email or offer to enter the realm name manually?

**Comment by darius-m** [comment]
I don't think that's possible and it would not really help, since clients reside in a specific realm. The client redirects the user to the SSO (if you exclude SAML IdP initiated logins) for authentication, so it must know which Keycloak realm it's associated to in order to redirect to the correct URL. A client can only be defined in a realm; even if the same name is reused in a different realm, it would be a different client with its own set of settings. Realms also differentiate sets of users - I don't think there are rules to disallow having users with the same email in different realms, so it would not be able to automatically infer the realm for that reason.

If you need to have multiple clients (from multiple realms) for a single application, you could probably add a list of buttons (or a dropdown and a button) to allow the user to select which realm they belong to in the application itself. Or perhaps you can define email-based filters in the application itself, like how Microsoft does it when delegating logins to an organization's SSO. It's similar to what you have requested, but it must be done in the application itself, and not in Keycloak.

What is your use-case for such functionality?

**Reply by voroninp** [reply]
Multitenancy.
Our clients are media agencies, and we want them to manage their users for our service.

**Reply by stringkey** [reply]
If you look at Okta for example there is realm discovery based on the domain part of the email that is used to identify the user. Based on this the realm is selected after the step where the user needs to enter their username. This is comon practice as many users don't even know what Idp they are authenticated against. So @voroninp his question is not that strange. Specifically when authenticating against azure active directory you dont want a list of buttons or dropdown exposing all your client idps to everybody. That would tell something about both your clientelle as the authentication machinism they use.

**Reply by stringkey** [reply]
Who can explain it better than an IdP solution that implemented this: https://auth0.com/docs/get-started/architecture-scenarios/business-to-business/authentication#hrd-through-universal-login-using-the-email-subdomain

**Reply by darius-m** [reply]
Yup, fair enough. I wasn't saying it was strange, just thought about it in a somewhat more limited way. I'm not aware that this would be possible in Keycloak out of the box, though.

I think you could get this result using a "common" realm, that provides the client for the application itself, and configure federations to the other realms (i.e., the application redirects to `/realms/common` for login, and from there you get redirected to `/realm/something` for the actual authentication). For this you would need to configure a client for "common" in all other realms, and then configure federations using OIDC / SAML for the other realms in "common". You can then configure the authentication flows in the "common" realm to redirect users to the correct federation using rules (this is the part I'm not sure is supported out of the box, but you can probably use an SPI to perform user-realm matching if it's not supported).

**Reply by darius-m** [reply]
Note that this is somewhat hackish and I'm not entirely sure how data in the "common" realm would look like without testing it myself. You'd probably need to make sure to handle some things with more care (e.g., what user data is stored in the realm and when it is updated - ideally it would not be stored in the "common" realm, or at the very least be updated on every login, all logins should go directly to the other realms to allow correct handling of session expiry and invalidation, which also means that no user sessions should be stored in "common", etc.).

**Reply by stringkey** [reply]
I think the strict separation between realms makes the implementation 'more' complex in KeyCloak. Okta and I think Auth0 (now effectively also Okta) have the concept of an 'all users' usergroup, this contains all users without any acreditions, the just are. Other groups can be created and user can be added to these groups making them similar to realms, the difference is I think that one or more applications (clients in IdP land) can be added in a many to many fashion. The realm discovey rule are setup per SAML2 federation trust and can live across multiple apps and groups.

**Reply by darius-m** [reply]
I think Keycloak may need to have this stricter separation, though. In a business to business setup, like Okta, Auth0, Microsoft (the enterprise part) you can count on organizations bringing their own email domains, so you cannot have users with the same email domains across multiple organizations.

Keycloak, however, supports user self-registration, so having duplicate email addresses is an inherent risk (if the same service is used by multiple organizations that allow self-registration). Automatic detection of the realm based on emails requires a strong guarantee that you don't have duplicate emails across the system. From a security point of view, it's not feasible to find which realms an email is registered into and present the list to the user before authentication to allow them to select a realm, since this would enable pretty easy account enumeration attacks.

**Comment by stringkey** [comment]
More then one user could share an email address but I don’t see how an email address can be non-unique. As far as I know there is a one to one relation between the email address and the identity.The only thing I can think off is that not the email address but another identifier is uniquely used to identify the user and the email is secondary. This is why you must use an email address and validate this email as the primary identifier. Email addresses are unique (when the RFC for case sensitivity is ignored as I think all mail servers do)I gladly hear why email addresses are not unique, maybe I have a gap in my knowledge or misunderstood what you meant.On 25 Jul 2024, at 20:28, darius-m ***@***.***> wrote:﻿
I think Keycloak may need to have this stricter separation, though. In a business to business setup, like Okta, Auth0, Microsoft (the enterprise part) you can count on organizations bringing their own email domains, so you cannot have users with the same email domains across multiple organizations.
Keycloak, however, supports user self-registration, so having duplicate email addresses is an inherent risk (if the same service is used by multiple organizations that allow self-registration). Automatic detection of the realm based on emails requires a strong guarantee that you don't have duplicate emails across the system. From a security point of view, it's not feasible to find which realms an email is registered into and present the list to the user before authentication to allow them to select a realm, since this would enable pretty easy account enumeration attacks.

**Reply by darius-m** [reply]
Users are uniquely identified by a UUID in the database, not the username or email. Validating the email is not required by default in the realm settings, and there's even an option to allow duplicate email addresses in the same realm. Even without the duplicate emails setting, I don't think there's any enforcement of email uniqueness across realms (as that would imply a user cannot create accounts for multiple businesses that use the same SSO, even though each businesses has its own realm).

**Comment by stringkey** [comment]
Like I said earlier, because of the segregation of realms you have the same user in multiple places. I understand that an identity is stored in the database with a unique id, that is developed domain. I was talking about the value that authenticates the user. As I understand it now the same user can exist in keycloak with unrelated identities per realm with the same email address. This is I think where the difference is with other IAM/IdP solutions. The same identity would be part of different groups, still being able to generate unique authentication token for different applications. But we are drifting in conversation from the original topic. As it currently stands there is no standard solution to do realm discovery with keycloak based on the domain part of the email address. Thanks for taking the time to respond. The philosophy behind it is a lot clearer to me now. On 25 Jul 2024, at 23:42, darius-m ***@***.***> wrote:﻿
Users are uniquely identified by a UUID in the database, not the username or email. Validating the email is not required by default in the realm settings, and there's even an option to allow duplicate email addresses in the same realm. Even without the duplicate emails setting, I don't think there's any enforcement of email uniqueness across realms (as that would imply a user cannot create accounts for multiple businesses that use the same SSO, even though each businesses has its own realm).
